ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -32,25 +33,100 @@ def __init__(self):
+
+    def loop(self):
+        rate = rospy.Rate(50)
+        while not rospy.is_shutdown():
+            if self.pose and self.base_waypoints:
+                closest_waypoint_idx = self.get_closest_waypoint_idx()
+                self.publish_waypoints(closest_waypoint_idx)
+                rate.sleep()
+
+    def get_closest_waypoint_idx(self):
+        x = self.pose.pose.position.x
+        y = self.pose.pose.position.y
+        closest_idx = self.waypoints_tree.query([x,y],1)[1]
+
+         # Check if closest is ahead or behind vehicle
+        closest_coord = self.waypoints_2d[closest_idx]
+        prev_coord = self.waypoints_2d[closest_idx -1]
+
+        # Equation for hyperplane through closest_coords
+        cl_vect = np.array(closest_coord)
+        prev_vect = np.array(prev_coord)
+        pos_vect = np.array([x,y])
+
+        val = np.dot(cl_vect-prev_vect, pos_vect-cl_vect)
+
+        if val > 0:
+            closest_idx = (closest_idx + 1)%len(self.waypoints_2d)
+        return closest_idx
+
+    def publish_waypoints(self):
+        final_lane = self.generate_lane()
+        self.final_waypoints_pub.publish(final_lane)
+
+    def generate_lane(self):
+        lane = Lane()
+
+        closest_idx = self.get_closest_waypoint_idx()
+        farthest_idx = closest_idx + LOOKAHEAD_WPS
+        base_waypoints = self.base_lane.waypoints[closest_idx:farthest_idx]
+
+        if (self.stopline_wp_idx == -1) or (self.stopline_wp_idx >= farthest_idx):
+            lane.waypoints = base_waypoints
+        else:
+            lane.waypoints = self.decelerate_waypoints(base_waypoints, closest_idx)
+
+        return lane
+
+    def decelerate_waypoints(self, waypoints, closest_idx):
+    temp = []
+    for i, wp in enumerate(waypoints):
+
+        p = Waypoint()
+        p.pose = wp.pose
+
+        # Distance includes a number of waypoints back so front of car stops at line
+        stop_idx = max(self.stopline_wp_idx - closest_idx - STOP_LINE_MARGIN, 0)
+        dist = self.distance(waypoints, i, stop_idx)
+        vel = math.sqrt(2 * MAX_DECEL * dist) + (i * CONSTANT_DECEL)
+        if vel < 1.0:
+            vel = 0.0
+
+        p.twist.twist.linear.x = min(vel, wp.twist.twist.linear.x)
+        temp.append(p)
+
+    self.decelerate_count += 1
+    if (self.decelerate_count % LOGGING_THROTTLE_FACTOR) == 0:
+        size = len(waypoints) - 1
+        vel_start = temp[0].twist.twist.linear.x
+        vel_end = temp[size].twist.twist.linear.x
+        rospy.logwarn("DECEL: vel[0]={:.2f}, vel[{}]={:.2f}".format(vel_start, size, vel_end))
+    return temp    
 
     def pose_cb(self, msg):
-        # TODO: Implement
-        pass
+        self.pose = msg
 
     def waypoints_cb(self, waypoints):
-        # TODO: Implement
-        pass
+        self.base_waypoints = waypoints
+        if not self.waypoints_2d:
+            self.waypoints_2d = [[Waypoint.pose.pose.position.x, Waypoint.pose.pose.position.y] for Waypoint.waypoint]
```

Comment:
There is a small syntax error here.
You could fix it like this (according to the video):
```
self.waypoints_2d = [[waypoint.pose.pose.position.x, waypoint.pose.pose.position.y] for waypoint in waypoints.waypoints]
```